<a href="https://colab.research.google.com/github/BhavyaSrinivas2003/DiabetesTestPredict/blob/main/DiabetesTestPrediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Diabetes prediction


In [ ]:
#importing standard libraries
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
#importing svm model
from sklearn import svm
from sklearn.metrics import accuracy_score

* Gathering Data

In [ ]:
diabetes_data=pd.read_csv("/content/diabetes.csv")
diabetes_data

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [ ]:
diabetes_data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [ ]:
diabetes_data.tail()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0

In [ ]:
diabetes_data.shape

(768, 9)

Checking Null Values

In [ ]:
diabetes_data.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

Checking wheather the data is balanced or unbalanced

In [ ]:
diabetes_data["Outcome"].value_counts()

0    500
1    268
Name: Outcome, dtype: int64

Hence the data is balanced in 2:1 ratio

* Splitting the data into 2 variables ---x & y
taking target function in y--variable and data in x--variable

Using drop method

In [ ]:
x=diabetes_data.drop("Outcome",axis=1)
x

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  
2                       0.672   32  
3                       0.167   21  
4                       2.288   33  
..                        ...  ...  
763                     0.171   63  
764                     0.340   27  
765                     0.245   30  
766                     0.349   47  
767                     0.315   23  

[768 rows x 8 columns]

In [ ]:
y=diabetes_data.Outcome
y

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64

Data Standardization

In [ ]:
scaler=StandardScaler()
scaler.fit(x)

StandardScaler()

In [ ]:
Standardised_data=scaler.transform(x)

In [ ]:
print(Standardised_data)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


Now taking this standardised data as x value because it has to be suitable for all the data

In [ ]:
x=Standardised_data

In [ ]:
print(x)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


In [ ]:
print(y)

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64


Splitting the data into train_data (80%) and test_data(20%) and taking shuffle_state nothing but random_state as 3

Train_Test_Split

In [ ]:
x_train, x_test, y_train, y_test= train_test_split(x,y,test_size=0.2,random_state=3)

In above test_size means 20% of test_data

Printing the shapes of train_data and test_data

In [ ]:
print(x.shape, x_train.shape, x_test.shape)

(768, 8) (614, 8) (154, 8)


In [ ]:
print(y.shape, y_train.shape, y_test.shape)

(768,) (614,) (154,)


Since y variable is target variable & it has only 1 column, we did'nt get any value after comma.

We use SVM model to make hyperplane between diabetic person(1) and non-diabetic person(0) and we can get high accuracy result.

Training the Model

In [ ]:
classifier=svm.SVC(kernel="linear")

SVC--Support Vector Classifier

In [ ]:
#train the support vector machine classifier
classifier.fit(x_train,y_train)

SVC(kernel='linear')

In [ ]:
print(classifier)

SVC(kernel='linear')


#Model Evaluation

* based on accuracy score

2 ways
* making prediction for the same data i.e test-data
* making prediction for new data

In [ ]:
#finding accuracy_score for the train_data
x_train_prediction= classifier.predict(x_train)
training_data_accuracy= accuracy_score(x_train_prediction, y_train)
print("Accuracy score of Train Data: ", training_data_accuracy)

Accuracy score of Train Data:  0.7850162866449512


the accuracy is only 78% beacause of weak data..Since we gave only 768 rows.

In [ ]:
#finding accuracy_score for the test_data
x_test_prediction=classifier.predict(x_test)
test_data_accuracy= accuracy_score(x_test_prediction, y_test)
print("Accuracy score of Test Data: ", test_data_accuracy)

Accuracy score of Test Data:  0.7337662337662337


Making Prediction

In [ ]:
input_data=[6, 148, 72, 35, 0, 33.6,0.627, 50 ]
#converting it into numpy array
input_data_as_numpy_array=np.array(input_data)
#reshaping the array since we are predicting
input_data_reshaped= input_data_as_numpy_array.reshape(1,-1)
#Standardise  the input data
std_data=scaler.transform(input_data_reshaped)
print(std_data)
prediction=classifier.predict(std_data)
print(prediction)
if prediction[0]==0:
  print("The person is not a diabetic")
elif prediction[0]==1:
  print("The person is diabetic")

[[ 0.63994726  0.84832379  0.14964075  0.90726993 -0.69289057  0.20401277
   0.46849198  1.4259954 ]]
[1]
The person is diabetic


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


Saving the trained Model

In [ ]:
import pickle

In [ ]:
filename="trained_model.sav"
pickle.dump(classifier,open(filename,'wb'))

* wb--write in binary form
* rb --read in binary form

In [ ]:
loaded_model=pickle.load(open('trained_model.sav','rb'))

In [ ]:
input_data=[6, 148, 72, 35, 0, 33.6,0.627, 50 ]
#converting it into numpy array
input_data_as_numpy_array=np.array(input_data)
#reshaping the array since we are predicting
input_data_reshaped= input_data_as_numpy_array.reshape(1,-1)
prediction=loaded_model.predict(input_data_reshaped)
print(prediction)
if prediction[0]==0:
  print("The person is not a diabetic")
elif prediction[0]==1:
  print("The person is diabetic")

[1]
The person is diabetic
